libs-haskell/da-hs-base/tests/Telemetry.hs

```diff
@@ -0,0 +1,75 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Telemetry(main) where
+
+import Control.Monad.Extra
+import DA.Daml.Project.Consts
+import DA.Service.Logger.Impl.GCP
+import DA.Service.Logger.Impl.Pure
+import DA.Test.Util
+import Data.Aeson (encode)
+import Data.Functor
+import qualified Data.Text as T
+import System.Directory
+import System.Environment.Blank
+import System.IO.Extra
+import Test.Tasty
+import Test.Tasty.HUnit
+
+main :: IO ()
+main =
+  withTempDir $ \tmpDir1 -> do
+  withTempDir $ \tmpDir2 -> do
+  setPermissions tmpDir1 emptyPermissions
+  withEnv [("TASTY_NUM_THREADS", Just "1")] $
+    defaultMain $ do
+      testGroup "Telemetry" [dataLimiterTests, cacheDirTests tmpDir1 tmpDir2]
+
+dataLimiterTests :: TestTree
+dataLimiterTests = testGroup "data limiter"
+    [ withResource homeDir (\mbHome -> whenJust mbHome unsetEnv) $ \getHomeDir ->
+      withResource (initialiseGcpState (GCPConfig "test" Nothing) makeNopHandle) (\gcp -> removeFile (sentDataFile gcp)) $ \getGcp ->
+          testCase "Check that limit is triggered" $ do
+             _ <- getHomeDir
+             gcp <- getGcp
+             let d = replicate 100 $ encode $ T.pack $ replicate 100000 'X'
+             res <- mapM (sendData gcp fakeSend) d
+             let l = length $ takeWhile isSuccess res
+             83 @=? l
+  ]
+
+cacheDirTests :: FilePath -> FilePath -> TestTree
+cacheDirTests badCacheDir cacheDir =
+    testGroup
+        "cache directory"
+        [ withResource (setEnv damlCacheEnvVar cacheDir True) (\() -> unsetEnv damlCacheEnvVar) $ \_ ->
+              withResource
+                  (initialiseGcpState (GCPConfig "test" (Just badCacheDir)) makeNopHandle)
+                  (\gcp -> removeFile (sentDataFile gcp)) $ \getGcp ->
+                  testCase
+                      "Check that opt-in/out, sendData, fetchMachineID works when a cache directory is specified." $ do
+                      gcp <- getGcp
+                      setOptIn gcp
+                      logOptOut gcp
+                      void $ fetchMachineID gcp
+                      let d = replicate 10 $ encode $ T.pack $ replicate 100 'X'
+                      res <- mapM (sendData gcp fakeSend) d
+                      let l = length $ takeWhile isSuccess res
+                      10 @=? l
+        ]
+
+fakeSend :: a -> IO ()
+fakeSend _ = pure ()
+
+-- | The CI env doesn't have a home directory so set and unset it if it doesn't exist
+homeDir :: IO (Maybe String)
+homeDir = do
+    home <- getEnv "HOME"
+    case home of
+        Nothing -> fakeHome
+        Just "" -> fakeHome
+        Just _ -> pure Nothing
+
+fakeHome :: IO (Maybe String)
```

Comment:
maybe set it to a temp directory instead?

Reply:
fixed.